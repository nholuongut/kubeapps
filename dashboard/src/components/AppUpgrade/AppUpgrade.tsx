import * as React from "react";

import { RouterAction } from "connected-react-router";
import { Link } from "react-router-dom";
import {
  IAppRepository,
  IChartState,
  IChartVersion,
  IRBACRole,
  IRelease,
} from "../../shared/types";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import LoadingWrapper from "../LoadingWrapper";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";

interface IAppUpgradeProps {
  app: IRelease;
  isFetching: boolean;
  error: Error | undefined;
  repoError: Error | undefined;
  kubeappsNamespace: string;
  namespace: string;
  releaseName: string;
  version: string;
  repos: IAppRepository[];
  repo: IAppRepository;
  selected: IChartState["selected"];
  upgradeApp: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
  ) => Promise<boolean>;
  clearRepo: () => void;
  checkChart: (repo: string, chartName: string) => void;
  fetchChartVersions: (id: string) => Promise<IChartVersion[]>;
  getAppWithUpdateInfo: (releaseName: string, namespace: string) => void;
  getChartVersion: (id: string, chartVersion: string) => void;
  getChartValues: (id: string, chartVersion: string) => void;
  push: (location: string) => RouterAction;
  goBack: () => RouterAction;
  fetchRepositories: () => void;
}

interface IAppUpgradeState {
  selectRepoForm: JSX.Element;
}

class AppUpgrade extends React.Component<IAppUpgradeProps, IAppUpgradeState> {
  public componentDidMount() {
    const { releaseName, getAppWithUpdateInfo, namespace, fetchRepositories } = this.props;
    getAppWithUpdateInfo(releaseName, namespace);
    fetchRepositories();
  }

  public render() {
    const { app, repos, error, namespace, releaseName, repoError, isFetching } = this.props;
    let { repo } = this.props;
    if (isFetching) {
      return <LoadingWrapper />;
    }
    if (
      !repos ||
      repos.length === 0 ||
      !app ||
      !app.chart ||
      !app.chart.metadata ||
      !app.chart.metadata.name ||
      !app.chart.metadata.version
    ) {
      if (error) {
        return (
          <ErrorSelector
            error={error}
            namespace={namespace}
            action="update"
            defaultRequiredRBACRoles={{ update: this.requiredRBACRoles() }}
            resource={releaseName}
          />
        );
      } else if (repoError) {
        return (
          <ErrorSelector
            error={repoError}
            namespace={this.props.kubeappsNamespace}
            action="view"
            defaultRequiredRBACRoles={{ view: this.requiredRBACRoles() }}
            resource="App Repositories"
          />
        );
      } else if (repos.length === 0) {
        return (
          <MessageAlert
            level={"warning"}
            children={
              <div>
                <h5>Chart repositories not found.</h5>
                Manage your Helm chart repositories in kubeapps by visiting the{" "}
                <Link to={"/config/repos"}>App repositories configuration</Link> page.
              </div>
            }
          />
        );
      } else {
        return (
          <ErrorSelector
            error={new Error("Unable to obtain the required information to upgrade")}
            resource={releaseName}
          />
        );
      }
    }
    // If there is update info about the app we can automatically chose the repository
    // with the latest version
    if (app.updateInfo) {
      const repoWithLatest = repos.find(
        r => r.metadata.name === (app.updateInfo && app.updateInfo.repository.name),
      );
      if (repoWithLatest) {
        repo = repoWithLatest;
      }
    }
    if (!repo.metadata) {
      return (
        <div>
          <SelectRepoForm
            {...this.props}
            error={this.props.repoError}
            chartName={app.chart.metadata.name}
          />
        </div>
      );
    }
    return (
      <div>
        <UpgradeForm
          {...this.props}
          appCurrentVersion={app.chart.metadata.version}
          appCurrentValues={(app.config && app.config.raw) || ""}
          chartName={app.chart.metadata.name}
          repo={repo.metadata.name}
        />
      </div>
    );
  }

  private requiredRBACRoles(): IRBACRole[] {
    return [
      {
        apiGroup: "kubeapps.com",
        namespace: this.props.kubeappsNamespace,
        resource: "apprepositories",
        verbs: ["get"],
      },
    ];
  }
}

export default AppUpgrade;
