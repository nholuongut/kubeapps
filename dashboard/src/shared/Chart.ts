import { axios } from "./AxiosInstance";
import { IChart } from "./types";

export default class Chart {
  public static async getReadme(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/README.md`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  public static async getValues(id: string, version: string) {
    const url = `${Chart.APIEndpoint}/assets/${id}/versions/${version}/values.yaml`;
    const { data } = await axios.get<string>(url);
    return data;
  }

  public static async exists(id: string, version: string, repo: string) {
    const url = `${Chart.APIEndpoint}/charts/${repo}/${id}/versions/${version}`;
    try {
      await axios.get<string>(url);
    } catch (e) {
      return false;
    }
    return true;
  }

  public static async listWithFilters(name: string, version: string, appVersion: string) {
    const url = `${
      Chart.APIEndpoint
    }/charts?name=${name}&version=${version}&appversion=${appVersion}`;
    const { data } = await axios.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/chartsvc/v1";
}
