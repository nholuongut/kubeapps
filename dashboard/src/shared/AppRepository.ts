import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getResourceLink(namespace?: string): string {
    return `${AppRepository.APIEndpoint}/${
      namespace ? `namespaces/${namespace}/` : ""
    }apprepositories`;
  }
  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories/${name}`;
  }
}
