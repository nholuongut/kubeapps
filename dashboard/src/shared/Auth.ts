import Axios, { AxiosResponse } from "axios";
const AuthTokenKey = "kubeapps_auth_token";
const AuthTokenOIDCKey = "kubeapps_auth_token_oidc";
import { APIBase } from "./Kube";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string, oidc: boolean) {
    localStorage.setItem(AuthTokenOIDCKey, oidc.toString());
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static usingOIDCToken() {
    const oidc = localStorage.getItem(AuthTokenOIDCKey);
    return oidc === "true" ? true : false;
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return [
      "base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, ""),
      "binary.k8s.io",
    ];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get(APIBase + "/", { headers: { Authorization: `Bearer ${token}` } });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      }
      // A 403 authorization error only occurs if the token resulted in
      // successful authentication. We don't make any assumptions over RBAC
      // for the root "/" nonResourceURL or other required authz permissions
      // until operations on those resources are attempted (though we may
      // want to revisit this in the future).
      if (res.status !== 403) {
        throw new Error(`${res.status}: ${res.data}`);
      }
    }
  }

  // fetchOIDCToken does a HEAD request to collect the Bearer token
  // from the authorization header if exists
  public static async fetchOIDCToken(): Promise<string | null> {
    try {
      const { headers } = await Axios.head("");
      if (headers && headers.authorization) {
        const tokenMatch = (headers.authorization as string).match(/Bearer\s(.*)/);
        if (tokenMatch) {
          return tokenMatch[1];
        }
      }
    } catch (e) {
      // Unable to retrieve token
    }
    return null;
  }
}
