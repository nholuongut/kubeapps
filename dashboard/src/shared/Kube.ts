import { Auth } from "./Auth";
import { axiosWithAuth } from "./AxiosInstance";
import { IResource } from "./types";

export const APIBase = "api/kube";
export let WebSocketAPIBase: string;
if (location.protocol === "https:") {
  WebSocketAPIBase = `wss://${window.location.host}${window.location.pathname}`;
} else {
  WebSocketAPIBase = `ws://${window.location.host}${window.location.pathname}`;
}

// We explicitly define the plurals here, just in case a generic pluralizer
// isn't sufficient. Note that CRDs can explicitly define pluralized forms,
// which might not match with the Kind. If this becomes difficult to
// maintain we can add a generic pluralizer and a way to override.
const ResourceKindToPlural = {
  Secret: "secrets",
  Service: "services",
  Ingress: "ingresses",
  Deployment: "deployments",
  StatefulSet: "statefulsets",
  DaemonSet: "daemonsets",
};

// Kube is a lower-level class for interacting with the Kubernetes API. Use
// ResourceRef to interact with a single API resource rather than using Kube
// directly.
export class Kube {
  public static getResourceURL(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let url = `${APIBase}/${apiVersion === "v1" ? "api/v1" : `apis/${apiVersion}`}`;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += `/${resource}`;
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static watchResourceURL(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let url = this.getResourceURL(apiVersion, resource, namespace);
    url = `${WebSocketAPIBase}${url}?watch=true`;
    if (name) {
      url += `&fieldSelector=metadata.name%3D${name}`;
    }
    if (query) {
      url += `&${query}`;
    }
    return url;
  }

  public static async getResource(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    const { data } = await axiosWithAuth.get<IResource>(
      this.getResourceURL(apiVersion, resource, namespace, name, query),
    );
    return data;
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public static watchResource(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    return new WebSocket(
      this.watchResourceURL(apiVersion, resource, namespace, name, query),
      Auth.wsProtocols(),
    );
  }

  // Gets the plural form of the resource Kind for use in the resource path
  public static resourcePlural(kind: string) {
    const plural = ResourceKindToPlural[kind];
    if (!plural) {
      throw new Error(`Don't know plural for ${kind}, register it in Kube`);
    }
    return plural;
  }
}
