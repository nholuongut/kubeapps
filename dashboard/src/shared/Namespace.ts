import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";

import { IK8sList, IResource } from "./types";

export default class Namespace {
  public static async list() {
    const { data } = await axiosWithAuth.get<IK8sList<IResource, {}>>(`${Namespace.APIEndpoint}`);
    return data;
  }

  private static APIBase: string = APIBase;
  private static APIEndpoint: string = `${Namespace.APIBase}/api/v1/namespaces`;
}

// Set of namespaces used accross the applications as default and "all ns" placeholders
export const definedNamespaces = {
  default: "default",
  all: "_all",
};
