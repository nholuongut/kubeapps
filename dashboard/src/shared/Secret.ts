import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { definedNamespaces } from "./Namespace";
import { IOwnerReference, ISecret } from "./types";

export default class Secret {
  public static async create(
    name: string,
    secrets: { [s: string]: string },
    owner: IOwnerReference | undefined,
    namespace: string = definedNamespaces.default,
  ) {
    const url = Secret.getLink(namespace);
    const { data } = await axiosWithAuth.post<ISecret>(url, {
      apiVersion: "v1",
      data: secrets,
      kind: "Secret",
      metadata: {
        name,
        ownerReferences: [owner],
      },
      type: "Opaque",
    });
    return data;
  }

  public static async delete(name: string, namespace: string = definedNamespaces.default) {
    const url = this.getLink(namespace, name);
    return axiosWithAuth.delete(url);
  }

  public static async get(name: string, namespace: string = definedNamespaces.default) {
    const url = this.getLink(namespace, name);
    const { data } = await axiosWithAuth.get<ISecret>(url);
    return data;
  }

  public static async list(namespace: string = definedNamespaces.default) {
    const url = Secret.getLink(namespace);
    const { data } = await axiosWithAuth.get<ISecret>(url);
    return data;
  }

  private static getLink(namespace: string, name?: string): string {
    return `${APIBase}/api/v1/namespaces/${namespace}/secrets${name ? `/${name}` : ""}`;
  }
}
