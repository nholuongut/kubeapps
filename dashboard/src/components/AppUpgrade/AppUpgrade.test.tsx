import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import { hapi } from "shared/hapi/release";
import itBehavesLike from "../../shared/specs";
import { ForbiddenError, IAppRepository, IChartState, IRelease } from "../../shared/types";
import { ErrorSelector, MessageAlert, PermissionsErrorAlert } from "../ErrorAlert";
import ErrorPageHeader from "../ErrorAlert/ErrorAlertHeader";
import SelectRepoForm from "../SelectRepoForm";
import UpgradeForm from "../UpgradeForm";
import AppUpgrade from "./AppUpgrade";

const defaultProps = {
  app: {} as hapi.release.Release,
  isFetching: false,
  checkChart: jest.fn(),
  clearRepo: jest.fn(),
  error: undefined,
  fetchChartVersions: jest.fn(),
  fetchRepositories: jest.fn(),
  getAppWithUpdateInfo: jest.fn(),
  getChartValues: jest.fn(),
  getChartVersion: jest.fn(),
  kubeappsNamespace: "kubeapps",
  namespace: "default",
  push: jest.fn(),
  goBack: jest.fn(),
  releaseName: "foo",
  repo: {} as IAppRepository,
  repoError: undefined,
  repos: [],
  selected: {} as IChartState["selected"],
  upgradeApp: jest.fn(),
  version: "1.0.0",
};

beforeEach(() => {
  jest.resetAllMocks();
});

itBehavesLike("aLoadingComponent", {
  component: AppUpgrade,
  props: { ...defaultProps, isFetching: true },
});

it("renders the repo selection form if not introduced", () => {
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as IRelease
      }
      repos={[
        {
          metadata: { name: "stable" },
        } as IAppRepository,
      ]}
    />,
  );
  expect(wrapper.find(SelectRepoForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(UpgradeForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

context("when an error exists", () => {
  it("renders a generic error message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = shallow(
      <AppUpgrade
        {...defaultProps}
        error={new Error("foo doesn't exists")}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.html()).toContain("Sorry! Something went wrong.");
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a forbidden message", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const role = {
      apiGroup: "kubeapps.com",
      namespace: "kubeapps",
      resource: "apprepositories",
      verbs: ["get"],
    };
    const wrapper = mount(
      <AppUpgrade
        {...defaultProps}
        error={new ForbiddenError(JSON.stringify([role]))}
        repos={[repo]}
        repo={repo}
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorPageHeader).text()).toContain(
      "You don't have sufficient permissions to update foo in the default namespace",
    );
    expect(wrapper.find(PermissionsErrorAlert).prop("roles")[0]).toMatchObject(role);
  });

  it("renders a forbidden message for the repositories", () => {
    const role = {
      apiGroup: "kubeapps.com",
      namespace: "kubeapps",
      resource: "apprepositories",
      verbs: ["get"],
    };
    const wrapper = mount(<AppUpgrade {...defaultProps} repoError={new ForbiddenError()} />);

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorPageHeader).text()).toContain(
      "You don't have sufficient permissions to view App Repositories in the kubeapps namespace",
    );
    expect(wrapper.find(PermissionsErrorAlert).prop("roles")[0]).toMatchObject(role);
  });

  it("renders a warning message if there are no repositories", () => {
    const wrapper = shallow(<AppUpgrade {...defaultProps} />);

    expect(wrapper.find(MessageAlert)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(
      wrapper
        .find(MessageAlert)
        .children()
        .text(),
    ).toContain("Chart repositories not found");
  });

  it("renders an error message if the app information is missing some metadata", () => {
    const repo = {
      metadata: { name: "stable" },
    } as IAppRepository;
    const wrapper = mount(
      <AppUpgrade
        {...defaultProps}
        repos={[repo]}
        app={
          {
            chart: {
              metadata: {},
            },
            name: "foo",
          } as IRelease
        }
      />,
    );

    expect(wrapper.find(ErrorSelector)).toExist();
    expect(wrapper.find(SelectRepoForm)).not.toExist();
    expect(wrapper.find(UpgradeForm)).not.toExist();

    expect(wrapper.find(ErrorSelector).text()).toContain(
      "Unable to obtain the required information to upgrade",
    );
  });
});

it("renders the upgrade form when the repo is available", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
        } as IRelease
      }
      repos={[repo]}
      repo={repo}
    />,
  );
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

it("skips the repo selection form if the app contains upgrade info", () => {
  const repo = {
    metadata: { name: "stable" },
  } as IAppRepository;
  const wrapper = shallow(
    <AppUpgrade
      {...defaultProps}
      app={
        {
          chart: {
            metadata: {
              name: "bar",
              version: "1.0.0",
            },
          },
          name: "foo",
          updateInfo: {
            upToDate: true,
            chartLatestVersion: "1.1.0",
            appLatestVersion: "1.1.0",
            repository: { name: "stable", url: "" },
          },
        } as IRelease
      }
      repos={[repo]}
      repo={repo}
    />,
  );
  expect(wrapper.find(UpgradeForm)).toExist();
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(SelectRepoForm)).not.toExist();
  expect(wrapper).toMatchSnapshot();
});
