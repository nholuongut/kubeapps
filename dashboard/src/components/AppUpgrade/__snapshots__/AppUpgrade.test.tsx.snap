// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the repo selection form if not introduced 1`] = `
<div>
  <SelectRepoForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
      }
    }
    chartName="bar"
    checkChart={[Function]}
    clearRepo={[Function]}
    fetchChartVersions={[Function]}
    fetchRepositories={[Function]}
    getAppWithUpdateInfo={[Function]}
    getChartValues={[Function]}
    getChartVersion={[Function]}
    goBack={[Function]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[Function]}
    releaseName="foo"
    repo={Object {}}
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[Function]}
    version="1.0.0"
  />
</div>
`;

exports[`renders the upgrade form when the repo is available 1`] = `
<div>
  <UpgradeForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
      }
    }
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    checkChart={[Function]}
    clearRepo={[Function]}
    fetchChartVersions={[Function]}
    fetchRepositories={[Function]}
    getAppWithUpdateInfo={[Function]}
    getChartValues={[Function]}
    getChartVersion={[Function]}
    goBack={[Function]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[Function]}
    releaseName="foo"
    repo="stable"
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[Function]}
    version="1.0.0"
  />
</div>
`;

exports[`skips the repo selection form if the app contains upgrade info 1`] = `
<div>
  <UpgradeForm
    app={
      Object {
        "chart": Object {
          "metadata": Object {
            "name": "bar",
            "version": "1.0.0",
          },
        },
        "name": "foo",
        "updateInfo": Object {
          "appLatestVersion": "1.1.0",
          "chartLatestVersion": "1.1.0",
          "repository": Object {
            "name": "stable",
            "url": "",
          },
          "upToDate": true,
        },
      }
    }
    appCurrentValues=""
    appCurrentVersion="1.0.0"
    chartName="bar"
    checkChart={[Function]}
    clearRepo={[Function]}
    fetchChartVersions={[Function]}
    fetchRepositories={[Function]}
    getAppWithUpdateInfo={[Function]}
    getChartValues={[Function]}
    getChartVersion={[Function]}
    goBack={[Function]}
    isFetching={false}
    kubeappsNamespace="kubeapps"
    namespace="default"
    push={[Function]}
    releaseName="foo"
    repo="stable"
    repos={
      Array [
        Object {
          "metadata": Object {
            "name": "stable",
          },
        },
      ]
    }
    selected={Object {}}
    upgradeApp={[Function]}
    version="1.0.0"
  />
</div>
`;

exports[`when an error exists renders a generic error message 1`] = `
<ErrorSelector
  action="update"
  defaultRequiredRBACRoles={
    Object {
      "update": Array [
        Object {
          "apiGroup": "kubeapps.com",
          "namespace": "kubeapps",
          "resource": "apprepositories",
          "verbs": Array [
            "get",
          ],
        },
      ],
    }
  }
  error={[Error: foo doesn't exists]}
  namespace="default"
  resource="foo"
/>
`;
